### PR TITLE
Use Custom Validator of c-r for webhook

### PR DESCRIPTION
**Pull request by acumino**
**How to categorize this PR?**

/area open-source
/kind enhancement

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Part of #5282

**Special notes for your reviewer**:
/cc @timebertt @rfranzke 

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go [hunk 1]
```diff
@@ -459,31 +459,319 @@ func GetValidatingWebhookConfig(caBundle []byte, webhookClientService *corev1.Se
 			SideEffects:             &sideEffect,
 			TimeoutSeconds:          pointer.Int32(10),
 		}, {
-			Name: "validation.extensions.seed.admission.core.gardener.cloud",
+			Name: "validation.extensions.backupbucket.admission.core.gardener.cloud",
```
Comment (rfranzke) [inline pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go]: Why do we now need one entry per resource?

File: pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go [hunk 2]
```diff
@@ -16,266 +16,290 @@ package extensionresources
...
-	// HandlerName is the name of this admission webhook handler.
-	HandlerName = "extension_resources"
+	// BackupBucketWebhookPath is the HTTP handler path for this admission webhook handler for BackupBucket.
+	BackupBucketWebhookPath = "/validate-extensions-gardener-cloud-v1alpha1-backupbucket"
+	// BackupEntryWebhookPath is the HTTP handler path for this admission webhook handler for BackupEntry.
+	BackupEntryWebhookPath = "/validate-extensions-gardener-cloud-v1alpha1-backupentry"
+	// BastionWebhookPath is the HTTP handler path for this admission webhook handler for Bastion.
+	BastionWebhookPath = "/validate-extensions-gardener-cloud-v1alpha1-bastion"
+	// ContaineRruntimeWebhookPath is the HTTP handler path for this admission webhook handler for ContainerRuntime.
+	ContaineRruntimeWebhookPath = "/validate-extensions-gardener-cloud-v1alpha1-containerruntime"
```
Comment (rfranzke) [inline pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: typo

File: pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go [hunk 3]
```diff
@@ -16,266 +16,290 @@ package extensionresources
...
-				return validation.ValidateWorker(n.(*extensionsv1alpha1.Worker))
-			},
-			validateUpdateResource: func(n, o client.Object) field.ErrorList {
-				return validation.ValidateWorkerUpdate(n.(*extensionsv1alpha1.Worker), o.(*extensionsv1alpha1.Worker))
-			},
-		},
+func AddWebhooks(mgr manager.Manager) error {
+	var err error
+	err = builder.WebhookManagedBy(mgr).WithValidator(&BackupBucketValidator{}).For(&extensionsv1alpha1.BackupBucket{}).Complete()
+	if err != nil {
```
Comment (rfranzke) [inline pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: Nit: Use this pattern everywhere:

```suggestion
	if err := builder.WebhookManagedBy(mgr).WithValidator(&BackupBucketValidator{}).For(&extensionsv1alpha1.BackupBucket{}).Complete(); err != nil {
```

File: pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go [hunk 4]
```diff
@@ -16,266 +16,290 @@ package extensionresources
...
+type BastionValidator struct{}
+type ContainerRuntimeValidator struct{}
+type ControlPlaneValidator struct{}
+type DNSRecordValidator struct{}
+type EtcdValidator struct{}
+type ExtensionValidator struct{}
+type InfrastructureValidator struct{}
+type NetworkValidator struct{}
+type OperatingSystemConfigValidator struct{}
+type WorkerValidator struct{}
```
Comment (rfranzke) [inline pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: Nit: Use `var ( ) ` and `struct ( )` blocks?

File: pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go [hunk 5]
```diff
@@ -16,266 +16,290 @@ package extensionresources
...
-		gvr("workers"): {
-			newObject: func() client.Object { return new(extensionsv1alpha1.Worker) },
-			validateCreateResource: func(n, _ client.Object) field.ErrorList {
-				return validation.ValidateWorker(n.(*extensionsv1alpha1.Worker))
-			},
-			validateUpdateResource: func(n, o client.Object) field.ErrorList {
-				return validation.ValidateWorkerUpdate(n.(*extensionsv1alpha1.Worker), o.(*extensionsv1alpha1.Worker))
-			},
-		},
+func AddWebhooks(mgr manager.Manager) error {
```
Comment (rfranzke) [inline pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go:61]: Doc strings for many exported functions in this file are missing

File: pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go (hunk 1 above)
Comment (acumino) [reply pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go]: In Client config, the `path` is different for every webhook.
```
ClientConfig: admissionregistrationv1.WebhookClientConfig{
				CABundle: caBundle,
				Service: &admissionregistrationv1.ServiceReference{
					Name:      webhookClientService.Name,
					Namespace: webhookClientService.Namespace,
					Path:      pointer.String(extensionresources.BackupBucketWebhookPath),
				},
			},
```

File: pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go (hunk 5 above)
Comment (acumino) [reply pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go:61]: This is the case for old commit it has been updated in this one c3569e03fa2d7113eae24dce5ff952dfcbc37dcc

File: pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go (hunk 1 above)
Comment (rfranzke) [reply pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go]: Ah, I see, makes sense, thank you!

File: test/integration/seedadmissioncontroller/seedadmission_suite_test.go [hunk 6]
```diff
@@ -89,9 +91,10 @@ var _ = BeforeSuite(func() {
 
 	By("setting up webhook server")
 	server := mgr.GetWebhookServer()
+	err = extensionresources.AddWebhooks(mgr)
+	Expect(err).NotTo(HaveOccurred())
```
Comment (rfranzke) [inline test/integration/seedadmissioncontroller/seedadmission_suite_test.go]:
```suggestion
	Expect(extensionresources.AddWebhooks(mgr)).To(Succeed())
```

File: pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go (hunk 4 above)
Comment (acumino) [reply pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: Not clear about struct block, can you explain what is needed? is it to create one more struct and wrap all struct in that?

Comment (rfranzke) [reply pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: Sorry, I meant `type ()` block, of course, just to improve readability a bit:

```golang
var (
    _ admission.CustomValidator = &BackupBucketValidator{}
    _ admission.CustomValidator = &BackupEntryValidator{}
    // ...
)

type (
    BackupBucketValidator struct{}
    BackupEntryValidator  struct{}
    // ...
)
```

**Comment by acumino** [comment]
/hold
until the `DenyInvalidExtesnionResource` feature gate is promoted to GA (#5793)

File: pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go [hunk 7]
```diff
@@ -16,266 +16,281 @@ package extensionresources
...
-			validateCreateResource: func(n, _ client.Object) field.ErrorList {
-				return validation.ValidateWorker(n.(*extensionsv1alpha1.Worker))
-			},
-			validateUpdateResource: func(n, o client.Object) field.ErrorList {
-				return validation.ValidateWorkerUpdate(n.(*extensionsv1alpha1.Worker), o.(*extensionsv1alpha1.Worker))
-			},
-		},
+// AddWebhooks add extension's valdation webhook to manager
+func AddWebhooks(mgr manager.Manager) error {
+	if err := builder.WebhookManagedBy(mgr).WithValidator(&backupBucketValidator{}).For(&extensionsv1alpha1.BackupBucket{}).Complete(); err != nil {
```
Comment (rfranzke) [inline pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: Another suggestion: What about something like this:

```go
for obj, validator := range map[client.Object]admission.CustomValidator{
    &extensionsv1alpha1.BackupBucket{}: &backupBucketValidator{},
    // ...
} {
    if err := builder.WebhookManagedBy(mgr).WithValidator(validator).For(obj).Complete(); err != nil {
        return err
    }
}
```

Comment (acumino) [reply pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: I kinda favor the current approach. But we can use this one also.

Comment (rfranzke) [reply pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: The current approach is very noisy and duplicates the same code again and again. Why do you favor it/what are the advantages?

Comment (acumino) [reply pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: Just one look gives the idea that many webhooks are being added not very strong very opinion.

Comment (acumino) [reply pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: Resolved in c0823b52f0a9ed4a2d6153de036df5de60edc962

**Review by rfranzke — COMMENTED** [review]
/lgtm
Any concerns @timebertt ?

**Comment by timebertt** [comment]
Sorry for the delay, will have a quick look.
/assign

File: pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go [hunk 8]
```diff
@@ -456,31 +456,319 @@ func GetValidatingWebhookConfig(caBundle []byte, webhookClientService *corev1.Se
...
```
Comment (timebertt) [inline pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go]: There is so much duplication going on here. Can you extract a function?

File: pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go [hunk 9]
```diff
@@ -16,261 +16,266 @@ package extensionresources
...
+	_ admission.CustomValidator = &bastionValidator{}
+	_ admission.CustomValidator = &containerRuntimeValidator{}
+	_ admission.CustomValidator = &controlPlaneValidator{}
+	_ admission.CustomValidator = &dnsRecordValidator{}
+	_ admission.CustomValidator = &etcdValidator{}
+	_ admission.CustomValidator = &extensionValidator{}
+	_ admission.CustomValidator = &infrastructureValidator{}
+	_ admission.CustomValidator = &networkValidator{}
+	_ admission.CustomValidator = &operatingSystemConfigValidator{}
+	_ admission.CustomValidator = &workerValidator{}
```
Comment (timebertt) [inline pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: You can drop these, no need to enforce interface compliance again, as it's already enforced by other calls and the map below.

File: pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go [hunk 10]
```diff
@@ -16,261 +16,266 @@ package extensionresources
...
+	dnsRecordValidator             struct{}
+	etcdValidator                  struct{}
+	extensionValidator             struct{}
+	infrastructureValidator        struct{}
+	networkValidator               struct{}
+	operatingSystemConfigValidator struct{}
+	workerValidator                struct{}
+)
+
+func (*backupBucketValidator) ValidateCreate(ctx context.Context, obj runtime.Object) error {
```
Comment (timebertt) [inline pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: Pointer receivers are not needed. Drop them everywhere.
```suggestion
func (backupBucketValidator) ValidateCreate(ctx context.Context, obj runtime.Object) error {
```

File: pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go [hunk 11]
```diff
@@ -16,261 +16,266 @@ package extensionresources
...
+	errors := validation.ValidateBackupBucket(obj.(*extensionsv1alpha1.BackupBucket))
+	return errors.ToAggregate()
```
Comment (timebertt) [inline pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go]: I would go for the following schema, which makes sure to correctly set `Invalid` as the response status reason:
```suggestion
	bb := obj.(*extensionsv1alpha1.BackupBucket)
	if errs := validation.ValidateBackupBucket(bb); len(errs) > 0 {
		return apierrors.NewInvalid(extensionsv1alpha1.Kind(extensionsv1alpha1.BackupBucketResource), bb.GetName(), errs)
	}
	return nil
```

See https://github.com/gardener/gardener/blob/c14f5d534e4e0a781bf6af0bb320da4e7a9c8e97/vendor/sigs.k8s.io/controller-runtime/pkg/webhook/admission/validator_custom.go#L101-L107

**Review by timebertt — CHANGES_REQUESTED** [review]

File: pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go [hunk 12]
```diff
@@ -456,7 +456,7 @@ func GetValidatingWebhookConfig(caBundle []byte, webhookClientService *corev1.Se
 			SideEffects:             &sideEffect,
 			TimeoutSeconds:          pointer.Int32(10),
 		}, {
-			Name: "validation.extensions.seed.admission.core.gardener.cloud",
+			Name: "validation.extensions.etcd.admission.core.gardener.cloud",
```
Comment (timebertt) [inline pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go:459]: Why is this one special and not contained in `getWebhooks`?

Comment (timebertt) [inline pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go:459]: Why is this one special and not contained in `getWebhooks`?

File: pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go [hunk 13]
```diff
@@ -466,22 +466,61 @@ func GetValidatingWebhookConfig(caBundle []byte, webhookClientService *corev1.Se
...
+				},
+			},
+			AdmissionReviewVersions: []string{admissionv1beta1.SchemeGroupVersion.Version, admissionv1.SchemeGroupVersion.Version},
+			MatchPolicy:             &matchPolicy,
+			SideEffects:             &sideEffect,
+			TimeoutSeconds:          pointer.Int32(10),
+		}},
+	}
+
+	webhookConfig.Webhooks = getWebhooks(caBundle, webhookClientService, webhookConfig.Webhooks)
```
Comment (timebertt) [inline pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go]: Can you do the `append` here?
```suggestion
	webhookConfig.Webhooks = append(webhookConfig.Webhooks, getWebhooks(caBundle, webhookClientService))
```

Comment (timebertt) [inline pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go:487]: Can you do the `append` here?
```suggestion
	webhookConfig.Webhooks = append(webhookConfig.Webhooks, getWebhooks(caBundle, webhookClientService))
```

File: pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go (hunk 12 above)
Comment (acumino) [reply pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go:459]: This is fine here, As it has many different things than others.

Comment (acumino) [reply pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go:459]: If we keep this also in `getwebhook` then we need to use `if else` in `APIGroup` and `APIVersion`, that kind of a little less readable.

**Review by timebertt — APPROVED** [review]
Thank you!
/lgtm

**Pull request merged**